host/windows/syscall.c

```diff
@@ -480,142 +948,469 @@ ssize_t oe_syscall_writev_ocall(
+    FindClose(pdir->hFind);
+    memset(&pdir->FindFileData, 0, sizeof(pdir->FindFileData));
+
+    pdir->hFind = FindFirstFileW(wpathname, &pdir->FindFileData);
+    if (pdir->hFind == INVALID_HANDLE_VALUE)
+    {
+        err = GetLastError();
+    }
+    pdir->dir_offs = 0;
 }
 
 int oe_syscall_closedir_ocall(uint64_t dirp)
 {
-    OE_UNUSED(dirp);
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
 
-    PANIC;
+    if (!dirp)
+    {
+        return -1;
+    }
+    if (!FindClose(pdir->hFind))
+    {
+        return -1;
+    }
+    free(pdir->pdirpath);
+    pdir->pdirpath = NULL;
+    free(pdir);
+    return 0;
 }
 
 int oe_syscall_stat_ocall(const char* pathname, struct oe_stat* buf)
 {
-    OE_UNUSED(pathname);
-    OE_UNUSED(buf);
+    int ret = -1;
+    WCHAR* wpathname = oe_syscall_path_to_win(pathname, NULL);
+    struct _stat64 winstat = {0};
 
-    PANIC;
+    ret = _wstat64(wpathname, &winstat);
+    if (ret < 0)
+    {
+        // How do we get to  wstat's error
+
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+
+#undef st_atime
+#undef st_mtime
+#undef st_ctime
+
+    buf->st_dev = winstat.st_dev;
+    buf->st_ino = winstat.st_ino;
+    buf->st_mode = win_stat_to_stat(winstat.st_mode);
+    buf->st_nlink = winstat.st_nlink;
+    buf->st_uid = winstat.st_uid;
+    buf->st_gid = winstat.st_gid;
+    buf->st_rdev = winstat.st_rdev;
+    buf->st_size = winstat.st_size;
+    buf->st_atim.tv_sec = winstat.st_atime;
+    buf->st_mtim.tv_sec = winstat.st_mtime;
+    buf->st_ctim.tv_sec = winstat.st_ctime;
+
+done:
+
+    if (wpathname)
+    {
+        free(wpathname);
+    }
+
+    return ret;
 }
 
 int oe_syscall_access_ocall(const char* pathname, int mode)
 {
-    OE_UNUSED(pathname);
-    OE_UNUSED(mode);
+    int ret = -1;
+    WCHAR* wpathname = oe_syscall_path_to_win(pathname, NULL);
 
-    PANIC;
+    int winmode = mode & ~1; // X_OK is a noop but makes access unhappy
+    ret = _waccess(wpathname, winmode);
+    if (ret < 0)
+    {
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+
+done:
+    if (wpathname)
+    {
+        free(wpathname);
+    }
+    return ret;
 }
 
 int oe_syscall_link_ocall(const char* oldpath, const char* newpath)
 {
-    OE_UNUSED(oldpath);
-    OE_UNUSED(newpath);
+    int ret = -1;
+    WCHAR* woldpath = oe_syscall_path_to_win(oldpath, NULL);
+    WCHAR* wnewpath = oe_syscall_path_to_win(newpath, NULL);
 
-    PANIC;
+    if (!CreateHardLinkW(wnewpath, woldpath, NULL))
+    {
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+    ret = 0;
+
+done:
+    if (woldpath)
+    {
+        free(woldpath);
+    }
+
+    if (wnewpath)
+    {
+        free(wnewpath);
+    }
+    return ret;
 }
 
 int oe_syscall_unlink_ocall(const char* pathname)
 {
-    OE_UNUSED(pathname);
+    int ret = -1;
+    WCHAR* wpathname = oe_syscall_path_to_win(pathname, NULL);
 
-    PANIC;
+    ret = _wunlink(wpathname);
```

Comment:
https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createhardlinkw
says use DeleteFile to delete hard links.

Reply:
Fixed, thanks!